server/src/main/java/org/elasticsearch/index/engine/ThreadPoolMergeQueue.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.index.engine;
+
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.concurrent.ConcurrentCollections;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.index.engine.ThreadPoolMergeScheduler.MergeTask;
+import org.elasticsearch.threadpool.ThreadPool;
+
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.PriorityBlockingQueue;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class ThreadPoolMergeQueue {
+    /**
+     * Floor for IO write rate limit (we will never go any lower than this)
+     */
+    private static final double MIN_MERGE_MB_PER_SEC = 5.0;
+    /**
+     * Ceiling for IO write rate limit (we will never go any higher than this)
+     */
+    private static final double MAX_MERGE_MB_PER_SEC = 10240.0;
+    /**
+     * Initial value for IO write rate limit when doAutoIOThrottle is true
+     */
+    private static final double START_MB_PER_SEC = 20.0;
+    private final AtomicInteger activeIOThrottledMergeTasksCount = new AtomicInteger();
+    private final PriorityBlockingQueue<MergeTask> queuedMergeTasks = new PriorityBlockingQueue<>();
+    // the set of all merge tasks currently being executed by merge threads from the pool,
+    // in order to be able to update the IO throttle rate of merge tasks also after they have started (while executing)
+    private final Set<MergeTask> currentlyRunningMergeTasks = ConcurrentCollections.newConcurrentSet();
+    /**
+     * Current IO write throttle rate, for all merge, across all merge schedulers (shards) on the node
+     */
+    private final AtomicReference<Double> targetMBPerSec = new AtomicReference<>(START_MB_PER_SEC);
```

Comment:
Should we make this an AtomicLong instead to avoid the boxing? We could simply use Bytes per sec instead, that should be plenty accuracy. Then we have to convert to a double upon change, which seems ok.

Reply:
As suggested, I've made the IO rate a `long` in 32dfaf86.